crates/btcio/src/reader/query.rs

```diff
@@ -279,7 +283,56 @@ async fn poll_for_new_blocks<R: Reader>(
         };
     }
 
-    Ok(events)
+    Ok(())
+}
+
+/// Watches the status channel for chainstate which has the finalized state matching as referred by
+/// checkpoint. If it is found, returns `true`, otherwise returns `false`. `false` is returned when
+/// there is a different chainstate than indicated by checkpoint for given slot.
+async fn wait_until_checkpoint_accepted_or_rejected<R: Reader>(
+    ctx: &ReaderContext<R>,
+    checkpoint: &SignedCheckpoint,
+) -> anyhow::Result<bool> {
+    let batch_info = checkpoint.checkpoint().batch_info();
+    let l2_commt = checkpoint.checkpoint().batch_info().l2_range.1;
+    let ckpt_epoch = batch_info.epoch();
+
+    let s = info_span!("wait_checkpoint_accepted", %ckpt_epoch, last_blkid=%l2_commt.blkid());
+    let _ = s.enter();
```

Comment:
Careful, `.enter()` interacts poorly with async code due to how future polling works.  The preferred way to do this is to wrap a future like `.instrument(span)`.

See: https://docs.rs/tracing/latest/tracing/span/struct.EnteredSpan.html#in-asynchronous-code

If you're unlucky then you end up entering the span many times and it breaks the logging output because you overflow buffers and spans go unclosed or something weird.

Reply:
Ohh, didn't know that. Thanks.

> (ask me how I know 🙃)

Okay I'll bite. How do you know? 🙃

Reply:
Well basically that.  And then the trace viewer I was using locked up for several seconds while updating the UI to display a trace that had several hundred (at least) nested spans, and then also displayed some error about a corrupted trace.